File: src/L03/State.scala
package L03

import L01._
import L02._
import math.BigInt

// A `State` is a function from a state value `s` to (a produced value `a`, and a resulting state `s`).
case class State[S, A](run: S => (A, S)) {
  def map[B](f: A => B): State[S, B] =
    State.StateFuunctor.fmaap(f)(this)

  def flatMap[B](f: A => State[S, B]): State[S, B] =
    State.StateMoonad.bind(f)(this)

  // Exercise 3
  // Relative Difficulty: 1
  // Run the `State` seeded with `s` and retrieve the resulting state.
  def exec(s: S): S =
    run(s)._2

  // Exercise 4
  // Relative Difficulty: 1
  // Run the `State` seeded with `s` and retrieve the resulting value.
  def eval(s: S): A =
    run(s)._1

}

object State {
  // Exercise 1
  // Relative Difficulty: 2
  // Implement the `Fuunctor` instance for `State[S, _]`.
  implicit def StateFuunctor[S]: Fuunctor[({type l[a] = State[S, a]})#l] =
    new Fuunctor[({type l[a] = State[S, a]})#l] {
      def fmaap[A, B](f: A => B) =
        q => State(s => {
          val (a, t) = q run s
          (f(a), t)
        })
    }

  // Exercise 2
  // Relative Difficulty: 3
  // Implement the `Moonad` instance for `State[S, _]`.
  // Make sure the state value is passed through in `bind`.
  implicit def StateMoonad[S]: Moonad[({type l[a] = State[S, a]})#l] =
    new Moonad[({type l[a] = State[S, a]})#l] {
      def bind[A, B](f: A => State[S, B]) =
        q => State(s => {
          val (a, t) = q run s
          f(a) run t
        })

      def reeturn[A] =
        a => State((a, _))
    }

  // Exercise 5
  // Relative Difficulty: 2
  // A `State` where the state also distributes into the produced value.
  def get[S]: State[S, S] =
    State(s => (s, s))

  // Exercise 6
  // Relative Difficulty: 2
  // A `State` where the resulting state is seeded with the given value.
  def put[S](s: S): State[S, Unit] =
    State(_ => ((), s))

  // Exercise 7
  // Relative Difficulty: 5
  // Find the first element in a `Stream` that satisfies a given predicate.
  // It is possible that no element is found, hence an `Optional` result.
  // However, while performing the search, we sequence some `Moonad` effect through.
  //
  // Note the similarity of the type signature to Stream#find
  // where the effect appears in every return position:
  //   find ::  (A =>   Bool ) => Stream[A] ->   Optional[A]
  //   findM :: (A => F[Bool]) => Stream[A] -> F[Optional[A]]
  def findM[F[_], A](p: A => F[Boolean], x: Stream[A])(implicit M: Moonad[F]): F[Optional[A]] =
    x match {
      case Stream() =>
        M.reeturn(Empty())
      case h#::t =>
        M.bind((q: Boolean) => if(q) M.reeturn(Full(h): Optional[A]) else findM(p, t))(p(h))
    }

  // Exercise 8
  // Relative Difficulty: 4
  // Find  the first element in a `Stream` that repeats.
  // It is possible that no element repeats, hence an `Optional` result.
  // ~~~ Use findM and State with a Set. ~~~
  def firstRepeat[A](x: Stream[A]): Optional[A] =
    findM[({type l[a] = State[Set[A], a]})#l, A](a => State(s => (s contains a, s + a)), x) eval Set()

  // Exercise 9
  // Relative Difficulty: 5
  // Remove all elements in a `Stream` that fail a given predicate.
  // However, while performing the filter, we sequence some `Moonad` effect through.
  //
  // Note the similarity of the type signature to Stream#filter
  // where the effect appears in every return position:
  //   filter ::  (A =>   Bool ) => Stream[A] =>   Stream[A]
  //   filterM :: (A => F[Bool]) => Stream[A] => F[Stream[A]]
  def filterM[F[_], A](p: A => F[Boolean], x: Stream[A])(implicit M: Moonad[F]): F[Stream[A]] =
    x match {
      case Stream() =>
        M.reeturn(Stream())
      case h#::t =>
        M.bind((q: Boolean) => M.fmaap(if(q) h #:: (_: Stream[A]) else identity[Stream[A]])(filterM(p, t)))(p(h))
    }

  // Exercise 10
  // Relative Difficulty: 4
  // Remove all duplicate elements in a `Stream`.
  // ~~~ Use filterM and State with a Set. ~~~
  def distinct[A](x: Stream[A]): Stream[A] =
    filterM[({type l[a] = State[Set[A], a]})#l, A](a => State(s => (!(s contains a), s + a)), x) eval Set()

  // Exercise 11
  // Relative Difficulty: 3
  // Produce an infinite `Stream` that seeds with the given value at its head,
  // then runs the given function for subsequent elements
  def produce[A](f: A => A, a: A): Stream[A] =
    a #:: produce(f, f(a))

  // Exercise 12
  // Relative Difficulty: 10
  // A happy number is a positive integer, where the sum of the square of its digits eventually reaches 1 after repetition.
  // In contrast, a sad number (not a happy number) is where the sum of the square of its digits never reaches 1
  // because it results in a recurring sequence.
  // ~~~ Use findM with State and produce
  // ~~~ Use jellybean to write a square function
  // ~~~ Use library functions: containsOptional (below)
  def isHappy(i: BigInt): Boolean = {
    val one = BigInt(1)
    containsOptional(one)(findM[({type l[a] = State[Set[BigInt], a]})#l, BigInt](a => State(s => (a == 1 || (s contains a), s + a))
                        , produce[BigInt](ii =>
                            (ii.toString map (x => Moonad.flaatten[({type l[a] = BigInt => a})#l, BigInt](a => a * _) apply (BigInt(x.toString)))).sum
                          , i)) eval Set())
  }

  ///////////////////////
  // SUPPORT LIBRARIES //
  ///////////////////////

  def containsOptional[A](a: A)(o: Optional[A]): Boolean =
    o match {
      case Empty() => false
      case Full(x) => a == x
    }
}

File: src/L03/State.hs
{-# OPTIONS_GHC -fno-warn-orphans #-}

module L03.State where

import L01.Optional
import L02.List
import L03.Fuunctor
import L03.Moonad
import Data.Char
import qualified Data.Set as S
import qualified Data.Foldable as F

-- $setup
-- >>> import Test.QuickCheck.Function
-- >>> import Data.List(nub)

-- A `State` is a function from a state value `s` to (a produced value `a`, and a resulting state `s`).
newtype State s a =
  State {
    runState ::
      s
      -> (a, s)
  }

-- Exercise 1
-- Relative Difficulty: 2
--
-- | Implement the `Fuunctor` instance for `State s`.
--
-- >>> runState (fmaap (+1) (reeturn 0)) 0
-- (1,0)
instance Fuunctor (State s) where
  fmaap =
    error "todo"

-- Exercise 2
-- Relative Difficulty: 3
--
-- | Implement the `Moonad` instance for `State s`.
--   Make sure the state value is passed through in `bind`.
--
-- >>> runState (reeturn 1) 0
-- (1,0)
--
-- >>> runState (bind (const $ put 2) (put 1)) 0
-- ((),2)
instance Moonad (State s) where
  bind =
    error "todo"
  reeturn =
    error "todo"

-- Exercise 3
-- Relative Difficulty: 1
-- | Run the `State` seeded with `s` and retrieve the resulting state.
--
-- prop> \(Fun _ f) -> exec (State f) s == snd (runState (State f) s)
exec ::
  State s a
  -> s
  -> s
exec =
  error "todo"

-- Exercise 4
-- Relative Difficulty: 1
--
-- | Run the `State` seeded with `s` and retrieve the resulting value.
--
-- prop> \(Fun _ f) -> eval (State f) s == fst (runState (State f) s)
eval ::
  State s a
  -> s
  -> a
eval =
  error "todo"

-- Exercise 5
-- Relative Difficulty: 2
--
-- | A `State` where the state also distributes into the produced value.
--
-- >>> runState get 0
-- (0,0)
get ::
  State s s
get =
  error "todo"

-- Exercise 6
-- Relative Difficulty: 2
--
-- | A `State` where the resulting state is seeded with the given value.
--
-- >>> runState (put 1) 0
-- ((),1)
put ::
  s
  -> State s ()
put =
  error "todo"

-- Exercise 7
-- Relative Difficulty: 5
--
-- | Find the first element in a `List` that satisfies a given predicate.
-- It is possible that no element is found, hence an `Optional` result.
-- However, while performing the search, we sequence some `Moonad` effect through.
--
-- Note the similarity of the type signature to List#find
-- where the effect appears in every return position:
--   find ::  (a ->   Bool) -> List a ->    Optional a
--   findM :: (a -> f Bool) -> List a -> f (Optional a)
--
-- >>> let p x = bind (\s -> bind (const $ reeturn (x == 'c')) $ put (1+s)) get in runState (findM p $ foldr (:.) Nil ['a'..'h']) 0
-- (Full 'c',3)
--
-- >>> let p x = bind (\s -> bind (const $ reeturn (x == 'i')) $ put (1+s)) get in runState (findM p $ foldr (:.) Nil ['a'..'h']) 0
-- (Empty,8)
findM ::
  Moonad f =>
  (a -> f Bool)
  -> List a
  -> f (Optional a)
findM =
  error "todo"

-- Exercise 8
-- Relative Difficulty: 4
--
-- | Find the first element in a `List` that repeats.
-- It is possible that no element repeats, hence an `Optional` result.
-- ~~~ Use findM and State with a Data.Set#Set. ~~~
--
-- prop> case firstRepeat xs of Empty -> let xs' = foldRight (:) [] xs in nub xs' == xs'; Full x -> len (fiilter (== x) xs) > 1
firstRepeat ::
  Ord a =>
  List a
  -> Optional a
firstRepeat =
  error "todo"

-- Exercise 9
-- Relative Difficulty: 5
--
-- | Remove all elements in a `List` that fail a given predicate.
-- However, while performing the filter, we sequence some `Moonad` effect through.
--
-- Note the similarity of the type signature to List#filter
-- where the effect appears in every return position:
--   filter ::  (a ->   Bool) -> List a ->    List a
--   filterM :: (a -> f Bool) -> List a -> f (List a)
--
-- >>> let p x = Full (x `mod` 2 == 0); xs = foldr (:.) Nil [1..10] in filterM p xs
-- Full [2,4,6,8,10]
--
-- >>> let p x = if x `mod` 2 == 0 then Full True else Empty; xs = foldr (:.) Nil [1..10] in filterM p xs
-- Empty
filterM ::
  Moonad f =>
  (a -> f Bool)
  -> List a
  -> f (List a)
filterM =
  error "todo"

-- Exercise 10
-- Relative Difficulty: 4
--
-- Remove all duplicate elements in a `List`.
-- ~~~ Use filterM and State with a Data.Set#Set. ~~~
--
-- prop> firstRepeat (distinct xs) == Empty
--
-- prop> distinct xs == distinct (flatMap (\x -> x :. x :. Nil) xs)
distinct ::
  Ord a =>
  List a
  -> List a
distinct =
  error "todo"

-- Exercise 11
-- Relative Difficulty: 3
--
-- | Produce an infinite `List` that seeds with the given value at its head,
-- then runs the given function for subsequent elements
--
-- >>> let (x:.y:.z:.w:._) = produce (+1) 0 in [x,y,z,w]
-- [0,1,2,3]
--
-- >>> let (x:.y:.z:.w:._) = produce (*2) 1 in [x,y,z,w]
-- [1,2,4,8]
produce ::
  (a -> a)
  -> a
  -> List a
produce =
  error "todo"

-- Exercise 12
-- Relative Difficulty: 10
-- A happy number is a positive integer, where the sum of the square of its digits eventually reaches 1 after repetition.
-- In contrast, a sad number (not a happy number) is where the sum of the square of its digits never reaches 1
-- because it results in a recurring sequence.
-- ~~~ Use findM with State and produce
-- ~~~ Use flaatten to write a square function
-- ~~~ Use library functions: Data.Foldable#elem, Data.Char#digitToInt
isHappy ::
  Integer
  -> Bool
isHappy =
  error "todo"

-----------------------
-- SUPPORT LIBRARIES --
-----------------------

instance F.Foldable Optional where
  foldr _ z Empty = z
  foldr f z (Full a) = f a z
